src/libraries/Common/src/Internal/Cryptography/UniversalCryptoDecryptor.cs

```diff
@@ -52,72 +52,111 @@ protected sealed override int UncheckedTransformBlock(byte[] inputBuffer, int in
                 }
 
                 // Postpone the last block to the next round.
-                Debug.Assert(inputCount >= _heldoverCipher.Length, "inputCount >= _heldoverCipher.Length");
-                int startOfLastBlock = inputOffset + inputCount - _heldoverCipher.Length;
-                Buffer.BlockCopy(inputBuffer, startOfLastBlock, _heldoverCipher, 0, _heldoverCipher.Length);
-                inputCount -= _heldoverCipher.Length;
-                Debug.Assert(inputCount % InputBlockSize == 0, "Did not remove whole blocks for depadding");
+                Debug.Assert(inputBuffer.Length >= _heldoverCipher.Length, "inputBuffer.Length >= _heldoverCipher.Length");
+                inputBuffer.Slice(inputBuffer.Length - _heldoverCipher.Length).CopyTo(_heldoverCipher);
+                inputBuffer = inputBuffer.Slice(0, inputBuffer.Length - _heldoverCipher.Length);
+                Debug.Assert(inputBuffer.Length % InputBlockSize == 0, "Did not remove whole blocks for depadding");
             }
 
-            if (inputCount > 0)
+            if (inputBuffer.Length > 0)
             {
-                decryptedBytes += BasicSymmetricCipher.Transform(inputBuffer, inputOffset, inputCount, outputBuffer, outputOffset);
+                decryptedBytes += BasicSymmetricCipher.Transform(inputBuffer, outputBuffer);
             }
 
             return decryptedBytes;
         }
 
-        protected sealed override byte[] UncheckedTransformFinalBlock(byte[] inputBuffer, int inputOffset, int inputCount)
+        protected override unsafe int UncheckedTransformFinalBlock(ReadOnlySpan<byte> inputBuffer, Span<byte> outputBuffer)
         {
             // We can't complete decryption on a partial block
-            if (inputCount % InputBlockSize != 0)
+            if (inputBuffer.Length % InputBlockSize != 0)
                 throw new CryptographicException(SR.Cryptography_PartialBlock);
 
             //
             // If we have postponed cipher bits from the prior round, copy that into the decryption buffer followed by the input data.
             // Otherwise the decryption buffer is just the input data.
             //
 
-            byte[]? ciphertext = null;
+            ReadOnlySpan<byte> inputCiphertext;
+            Span<byte> ciphertext;
 
             if (_heldoverCipher == null)
             {
-                ciphertext = new byte[inputCount];
-                Buffer.BlockCopy(inputBuffer, inputOffset, ciphertext, 0, inputCount);
+#if NETSTANDARD || NETFRAMEWORK || NETCOREAPP3_0
+                ciphertext = new byte[inputBuffer.Length];
+#else
+                ciphertext = GC.AllocateUninitializedArray<byte>(inputBuffer.Length);
```

Comment:
We're not returning these arrays or calling back into callbacks, perhaps CryptoPool renting would be better?